gluten-ut/spark33/src/test/scala/org/apache/spark/sql/execution/datasources/GlutenFileMetadataStructSuite.scala

```diff
@@ -16,6 +16,156 @@
+              df0.coalesce(1).write.format(testFileFormat).save(f0)
+
+              val df1 = spark.createDataFrame(data1.asJava, fileSchema)
+              val f1 = new File(dir, "data/f1 gluten").getCanonicalPath
+              df1.coalesce(1).write.format(testFileFormat).save(f1)
+
+              // 2. read both f0 and f1
+              val df = spark.read
+                .format(testFileFormat)
+                .schema(fileSchema)
+                .load(new File(dir, "data").getCanonicalPath + "/*")
+              val realF0 = new File(dir, "data/f0")
+                .listFiles()
+                .filter(_.getName.endsWith(s".$testFileFormat"))
+                .head
+              val realF1 = new File(dir, "data/f1 gluten")
+                .listFiles()
+                .filter(_.getName.endsWith(s".$testFileFormat"))
+                .head
+              f(df, getMetadataForFile(realF0), getMetadataForFile(realF1))
+          }
+        }
+    }
+  }
+
+  metadataColumnsNativeTest(
+    "plan check with metadata and user data select",
+    schemaWithFilePathField) {
+    (df, f0, f1) =>
+      var dfWithMetadata = df.select(
+        METADATA_FILE_NAME,
+        METADATA_FILE_PATH,
+        METADATA_FILE_SIZE,
+        METADATA_FILE_MODIFICATION_TIME,
+        "age")
+      dfWithMetadata.collect
+      var fileScan = dfWithMetadata.queryExecution.executedPlan.collect {
+        case f: FileSourceScanExec => f
+      }
+      assert(fileScan.size == 1)
+      if (BackendTestUtils.isVeloxBackendLoaded()) {
+        assert(fileScan(0).nodeNamePrefix == "NativeFile")
+      } else {
+        assert(fileScan(0).nodeNamePrefix == "File")
+      }
+
+      // would fallback
+      dfWithMetadata = df.select(METADATA_FILE_PATH, "file_path")
+      checkAnswer(
+        dfWithMetadata,
+        Seq(
+          Row(f0(METADATA_FILE_PATH), "jack"),
+          Row(f1(METADATA_FILE_PATH), "lily")
+        )
+      )
+      fileScan = dfWithMetadata.queryExecution.executedPlan.collect {
+        case f: FileSourceScanExec => f
+      }
+      assert(fileScan(0).nodeNamePrefix == "File")
+  }
+
+  metadataColumnsNativeTest("plan check with metadata filter", schemaWithFilePathField) {
+    (df, f0, f1) =>
+      var filterDF = df
+        .select("file_path", "age", METADATA_FILE_NAME)
+        .where(Column(METADATA_FILE_NAME) === f0((METADATA_FILE_NAME)))
+      val ret = filterDF.collect
+      assert(ret.size == 1)
+      var fileScan = filterDF.queryExecution.executedPlan.collect {
+        case f: FileSourceScanExec => f
+      }
+      assert(fileScan.size == 1)
+      if (BackendTestUtils.isVeloxBackendLoaded()) {
+        assert(fileScan(0).nodeNamePrefix == "NativeFile")
+      } else {
+        assert(fileScan(0).nodeNamePrefix == "File")
+      }
+      var filterExecs = filterDF.queryExecution.executedPlan.collect {
+        case filter: FilterExecTransformerBase => filter
+      }
+      assert(filterExecs.size == 1)
```

Comment:
the goal is to check the scan is executed natively. can we use similar approach: 
https://github.com/apache/incubator-gluten/blob/main/backends-velox/src/test/scala/io/glutenproject/execution/TestOperator.scala#L467

Reply:
+1